src/txn/txn_rollback_nondurable.c

```diff
@@ -0,0 +1,319 @@
+		    "requires a timestamp greater than the pinned timestamp");
+
+	/*
+	 * Help the user - see if they have any active transactions. I'd
+	 * like to check the transaction running flag, but that would
+	 * require peeking into all open sessions, which isn't really
+	 * kosher.
+	 */
+	WT_ORDERED_READ(session_cnt, S2C(session)->session_cnt);
+	for (i = 0, s = txn_global->states; i < session_cnt; i++, s++) {
+		if (s->id != WT_TXN_NONE || s->pinned_id != WT_TXN_NONE)
+			WT_RET_MSG(session, EINVAL,
+			    "rollback_nondurable_commits not supported with "
+			    "active transactions");
+	}
+
+	return (0);
+}
+
+/*
+ * __txn_abort_newer_update --
+ *	Review an update chain
+ */
+static void
+__txn_abort_newer_update(WT_SESSION_IMPL *session,
+    WT_UPDATE *upd, wt_timestamp_t rollback_timestamp)
+{
+	bool aborted_one = false;
+	WT_UPDATE *next_upd;
+
+	for (next_upd = upd; next_upd != NULL; next_upd = next_upd->next) {
+		/* Only updates with timestamps will be aborted */
+		if (!__wt_timestamp_iszero(next_upd->timestamp) &&
+		    __wt_timestamp_cmp(
+		    rollback_timestamp, next_upd->timestamp) < 0) {
+			next_upd->txnid = WT_TXN_ABORTED;
+			/* TODO: should we clear the timestamp? */
+
+			/*
+			* If any updates are aborted, all newer updates
+			* better be aborted as well.
+			*/
+			if (!aborted_one)
+				WT_ASSERT(session,
+				    !aborted_one || upd == next_upd);
+			aborted_one = true;
+		}
+	}
+}
+
+/*
+ * __txn_abort_newer_row_skip --
+ *	Apply the update abort check to each entry in an insert skip list
+ */
+static void
+__txn_abort_newer_row_skip(WT_SESSION_IMPL *session,
+    WT_INSERT_HEAD *head, wt_timestamp_t rollback_timestamp)
+{
+	WT_INSERT *ins;
+
+	WT_SKIP_FOREACH(ins, head)
+		__txn_abort_newer_update(session, ins->upd, rollback_timestamp);
+}
+
+/*
+ * __txn_abort_newer_row_leaf --
+ *	Abort updates on a row leaf page with timestamps too new.
+ */
+static void
+__txn_abort_newer_row_leaf(
+    WT_SESSION_IMPL *session, WT_PAGE *page, wt_timestamp_t rollback_timestamp)
+{
+	WT_INSERT_HEAD *insert;
+	WT_ROW *rip;
+	WT_UPDATE *upd;
+	uint32_t i;
+
+	/*
+	 * Review the insert list for keys before the first entry on the disk
+	 * page.
+	 */
+	if ((insert = WT_ROW_INSERT_SMALLEST(page)) != NULL)
+		__txn_abort_newer_row_skip(
+		    session, insert, rollback_timestamp);
+
+	/*
+	 * Review updates that belong to keys that are on the disk image,
+	 * as well as for keys inserted since the page was read from disk.
+	 */
+	WT_ROW_FOREACH(page, rip, i) {
+		if ((upd = WT_ROW_UPDATE(page, rip)) != NULL)
+			__txn_abort_newer_update(
+			    session, upd, rollback_timestamp);
+
+		if ((insert = WT_ROW_INSERT(page, rip)) != NULL)
+			__txn_abort_newer_row_skip(
+			    session, insert, rollback_timestamp);
+	}
+}
+
+/*
+ * __txn_abort_newer_fast_deleted --
+ *	Fixup deleted refs so they are left in the correct state
+ */
+static void
+__txn_abort_newer_fast_deleted(WT_REF *ref, wt_timestamp_t rollback_timestamp)
```

Comment:
Edit: I've stared at this now, and I remember the deal.

The `WT_REF` has to have a transaction ID because that's the only way we can know where the fast-delete operation appears in concurrency control. We could push the `WT_UPDATE` list into the transaction structure, but it doesn't simplify anything (and adds an additional lock cycle to instantiate a fast-delete page, which maybe doesn't matter since that should be a rare operation).

We could get rid of the `WT_PAGE_DELETED` structure by converting the `WT_UPDATE` list it contains (which is just an array of pointers), into a linked-list of `WT_UPDATE` structures, but *shrug*.